core/ir/aarch64/instr_create.h

```diff
@@ -539,6 +539,12 @@
 #define INSTR_CREATE_ldp(dc, rt1, rt2, mem) \
     instr_create_2dst_1src(dc, OP_ldp, rt1, rt2, mem)
 #define INSTR_CREATE_ldr(dc, Rd, mem) instr_create_1dst_1src((dc), OP_ldr, (Rd), (mem))
+#define INSTR_CREATE_ldr_imm(dc, Rt, Rn, imm, post)                                      \
```

Comment:
Not sure why this needs the "post": that is just an encoding detail that doesn't affect semantics, and it only applies to a zero offset, right?

Reply:
My comment is based on only passing this to opnd_create_far_base_disp_ex().  See the comment for the pre_index field in opnd.h.  For creating semantically-different variants of the instruction itself, those are going to change the other operands too and not just the memop, right?  Not sure what is currently done for other A64 instrs, but for A32 we have different variants of the macro itself for different write-back options.

Reply:
I should've read the explanation of the "pre_index" bitfield. Thank you.
I did gdb tracing and found pre_index is hard-coded to true, so I figured the way to workaround it is to post-process it (manually changing to false, after encoded).

Today, I've tried to passing a negative offset to opnd_create_far_base_disp_ex() and I found it doesn't encode correctly.
Report: Internal Error: DynamoRIO debug check failure: /home/merck/Code/dynamorio_orig/core/arch/emit_utils_shared.c:1061 nxt_pc != NULL

** I will look into it tomorrow to see what actually happens, but I guess the AArch64 encoder doesn't catch it.

I realized that when a pre_index LDR (immediate) is encoded the instr_t will be handled by encode_opndsgen_38400400() in encode_gen.h (line: 8085). The AArch64 encoder tries each possible encoding functions until one of those catches it.

LDR (immediate) differs pre- or post-index by looking at bit 11 and 12.
When it comes to encoding a post-index instruction, the encoder will fall through to next encode_opndsgen_38400c00() function (bit11 = 1), and the instr_t is supposed to be caught by it.
In the first commit of this pull request, I found toggling the pre_index bitfield does affect how AArch64 encodes LDR (immediate). But passing a negative offset to opnd_create_far_base_disp_ex() doesn't seem be encoded correctly in AArch64.

Let me have a look tmr, and I will come back with a better change for reviews. Thank you!

Reply:
BTW, thank you for laying out the design principles.

Reply:
> Today, I've tried passing a negative offset to opnd_create_far_base_disp_ex() and I found it doesn't encode correctly.
> Report: Internal Error: DynamoRIO debug check failure: /home/merck/Code/dynamorio_orig/core/arch/emit_utils_shared.c:1061 nxt_pc != NULL

I remember something off in the A64 branch linking code: #4273, but that's a different assert.  @AssadHashmi might know more?

Reply:
I've done some investigations about the post-LDR encoding problem.

Summary:
0) I think the explanation of pre_index bit-field in the opnd.h is described in terms of A64 decoding.
   For recording a decoded LDR is pre- or post-index, it uses the pre_index bit-field to distinguish it.
   The value of the pre_index will be written by A64 decoder according to the input app's code sequence,
   and the JIT is supposed to emit a correct LDR version by looking at the value of the pre_index field.
   
   It's not meant to say a post-index LDR cannot have an offset other than 0, but both pre- and post-index
   LDR could both have a zero offset encoded. So the pre_index field is needed for distinguishing them in DR.

   Therefore, for *manually* encoding an LDR in the DR program itself, the pre_index field is hard-coded to *true*.
   So if a developer calls OPND_CREATE_MEMPTR()->opnd_create_far_base_disp_ex() in DR.
   The JIT will never have a chance to emit a correct post-LDR instr.

Tl;Dr
1) Only LDR/STR family uses encode_opnd_mem9/encode_opnd_mem9post/encode_opnd_mem9_bytes functions to encode the signed offset imm9 field (bit 20:12). If a change happens in those will only affect LDR/STR family.

2) In the encode_opnd_mem9post() and the following encode_opnd_mem9_bytes() function, if post==true, then the disp is requested to be 0, otherwise it will fail the encoding (codec.c, line: 662). *This is where it fails.*
    if (post ? (disp != 0) : (disp < -256 || disp > 255))
        return false;

3) Based on point #1, regardless of the indexing direction, per A64 SPEC. both pre- and post-index LDR/STR support encoding a signed 0 imm9 offset (AARM, p. 1000). But in the point #2, current impl. limits an input of post==true to have a zero disp and allows post==false to have a disp of range from -256 to +255. *This impl. is contradicting with the A64 SPEC.*

4) Because the imm9 field's type is a signed integer. If passing an negative offset to opnd_create_far_base_disp_ex() function for the purpose of determining the pre- or post-index attribute of an LDR/STR encoding will result in prohibiting negative offsets to be correctly encoded in the further pre-index LDR/STR.
*Therefore, will not go this way, but will think about setting FLAG_NEGATIVE and converting value to positive*

5) The later AArch64 instruction encoding process doesn't determine a pre- or post-index LDR by directly looking at the pre_index bit-field. Instead, the A64 encoder tries each possible LDR encoding functions until one of those reports successful. *A bit inefficient in A64 encoding process.* Furthermore, if a preceding encoding function returns TRUE, it may cascadingly block a later *more correct* one. This actually complicates the constructions of opnd_t variables, because the relationship among those opnd_t variables determines the selection of the later A64 encoding functions. Rather than simply determining by looking at the values of individually opnd_t vars at first.

Conclusion:
6) I will study more about ARM's implementation (NEGATIVE FLAG, positive conversion), before I implement a change for A64. Will come back with a more considerate commit that is more aligned with current ARM port.

Reply:
Reverted the changes of opnd_create_far_base_disp_ex(). Refined INSTR_CREATE_ldr_imm() macros to align with DR design (moved away from AArch64 mnemonics style to DR style). If the developers want a post-indexed LDR, they are supposed to call opnd_set_zero_offset_pre_index() on the just generated memop to set the pre_index field from true to false (indicating post-indexed), see an example in core/arch/aarch64/emit_utils.c.